File: questao_1/java/dados/FonteArquivo.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.List;
import java.io.IOException;

public class FonteArquivo extends FonteDeDados {
  private List<String> palavras;
  private int index;

  public static void main(String[] args){
    FonteArquivo fonteArquivo = new FonteArquivo("db.txt");
    while(fonteArquivo.possuiDado()){
      System.out.println(fonteArquivo.proximoDado());
    }
  }
  public FonteArquivo(String endereco) {
    super(endereco);
    this.palavras = new ArrayList<String>();
    this.index = 0;
    try{
    BufferedReader bf = new BufferedReader(new FileReader(endereco));
    String linha;
    while((linha = bf.readLine()) != null){
      String[] palavrasLinha = linha.split("\\s+");
      for(String palavra : palavrasLinha){
        this.palavras.add(palavra);
      }
    }
    }catch (IOException e){
      e.printStackTrace();
    }
  }
  public boolean possuiDado(){
   return this.index < this.palavras.size();
  }
  public String proximoDado(){
    String data = this.palavras.get(this.index);
    this.index++;
    return data;
  }
}

File: questao_1/node/FonteArquivo.js
import FonteDeDados from "./FontesDeDados.js";
import fs from "fs/promises";

class FonteArquivo extends FonteDeDados {
  constructor(endereco) {
    super();
    this.endereco = endereco;
    this.index = 0;
    this.data = [];
  }
  async init() {
    try {
      const data = await fs.readFile(this.endereco, "utf8");
      this.data = data.split(" ");
    } catch (err) {
      console.log(err);
    }
  }
  possuiDado() {
    return this.index < this.data.length;
  }
  proximoDado() {
    const nextData = this.data[this.index];
    this.index++;
    return nextData;
  }
}

export default FonteArquivo;
